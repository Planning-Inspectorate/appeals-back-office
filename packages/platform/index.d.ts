import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
