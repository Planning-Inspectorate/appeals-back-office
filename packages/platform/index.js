export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
