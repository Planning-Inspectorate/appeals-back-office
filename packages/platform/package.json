{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
