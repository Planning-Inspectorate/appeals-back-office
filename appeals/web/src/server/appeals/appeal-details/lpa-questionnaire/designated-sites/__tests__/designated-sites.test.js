// @ts-nocheck
import {
	appealDataFullPlanning,
	designatedSiteNames,
	lpaDesignatedSites,
	lpaQuestionnaireDataNotValidated
} from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('in-near-or-likely-to-affect-designated-sites', () => {
	beforeEach(() => {
		installMockApi();
		nock('http://test/')
			.get('/appeals/2')
			.reply(200, {
				...appealDataFullPlanning,
				appealId: 2
			});
		nock('http://test/').get('/appeals/lpa-designated-sites').reply(200, lpaDesignatedSites);
	});
	afterEach(teardown);

	describe('GET /in-near-or-likely-to-affect-designated-sites/change', () => {
		it('should render the change in, near or likely to affect designated sites page with the expected content', async () => {
			nock('http://test/')
				.get('/appeals/2/lpa-questionnaires/1')
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					lpaQuestionnaireId: 1,
					designatedSiteNames: []
				});

			const response = await request.get(
				`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
			);

			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain(
				'Change whether the site is in, partly in, or likely to affect a sensitive area</h1>'
			);

			for (const lpaDesignatedSite of lpaDesignatedSites) {
				expect(unprettifiedHtml).toContain(
					`name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox" value="${lpaDesignatedSite.key}">`
				);
				expect(unprettifiedHtml).toContain(`${lpaDesignatedSite.name}</label>`);
			}

			expect(unprettifiedHtml).toContain(
				'name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox" value="other"'
			);
			expect(unprettifiedHtml).toContain(
				'for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-6"> Other</label>'
			);
			expect(unprettifiedHtml).toContain(
				'for="custom-designation" class="govuk-label">Other designation(s)</label>'
			);
			expect(unprettifiedHtml).toContain(
				'id="custom-designation" name="customDesignation" type="text"'
			);
			expect(unprettifiedHtml).toContain('class="govuk-checkboxes__divider">or</div>');
			expect(unprettifiedHtml).toContain(
				'name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox" value="none" checked data-behaviour="exclusive"'
			);
			expect(unprettifiedHtml).toContain(
				'for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-8"> No, it is not in, near or likely to affect any designated sites</label>'
			);
			expect(unprettifiedHtml).toContain(
				'<button type="submit" data-prevent-double-click="true" class="govuk-button" data-module="govuk-button"> Continue</button>'
			);
		});

		it('should render the change in, near or likely to affect designated sites page with the expected checkboxes checked if there are existing designatedSiteNames items in the LPA questionnaire', async () => {
			nock('http://test/')
				.get('/appeals/2/lpa-questionnaires/1')
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					lpaQuestionnaireId: 1,
					designatedSiteNames
				});

			const response = await request.get(
				`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
			);

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('type="checkbox" value="SSSI" checked>');
		});

		it('should render the change in, near or likely to affect designated sites page with the "Other" checkbox checked, and the corresponding custom designatedSite text populated in the conditional text input, if there is an existing custom designatedSiteNames item in the LPA questionnaire', async () => {
			nock('http://test/')
				.get('/appeals/2/lpa-questionnaires/1')
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					lpaQuestionnaireId: 1,
					designatedSiteNames
				});

			const response = await request.get(
				`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
			);

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('type="checkbox" value="other" checked');
			expect(unprettifiedHtml).toContain(
				'id="custom-designation" name="customDesignation" type="text" value="test custom designation">'
			);
		});

		it('should render the change in, near or likely to affect designated sites page with the "None" checkbox checked if there are no existing designatedSiteNames items in the LPA questionnaire', async () => {
			nock('http://test/')
				.get('/appeals/2/lpa-questionnaires/1')
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					lpaQuestionnaireId: 1,
					designatedSiteNames: []
				});

			const response = await request.get(
				`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
			);

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain(
				'type="checkbox" value="none" checked data-behaviour="exclusive">'
			);
		});
	});

	describe('POST /in-near-or-likely-to-affect-designated-sites/change', () => {
		beforeEach(() => {
			nock('http://test/')
				.get('/appeals/2/lpa-questionnaires/1')
				.reply(200, {
					...lpaQuestionnaireDataNotValidated,
					lpaQuestionnaireId: 1,
					designatedSiteNames: []
				});
		});

		it('should call the PATCH LPA questionnaire endpoint with the expected payload and redirect to the LPA Questionnaire page if a single checkbox is checked', async () => {
			const mockedLPAQPatchEndpoint = nock('http://test/')
				.patch('/appeals/2/lpa-questionnaires/1')
				.reply(200, {});

			const response = await request
				.post(
					`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
				)
				.send({
					inNearOrLikelyToAffectDesignatedSitesCheckboxes: 'SSSI',
					customDesignation: ''
				});

			expect(mockedLPAQPatchEndpoint.isDone()).toBe(true);
			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/2/lpa-questionnaire/1'
			);
		});

		it('should call the PATCH LPA questionnaire endpoint with the expected payload and redirect to the LPA Questionnaire page if multiple checkboxes are checked', async () => {
			const mockedLPAQPatchEndpoint = nock('http://test/')
				.patch('/appeals/2/lpa-questionnaires/1')
				.reply(200, {});

			const response = await request
				.post(
					`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
				)
				.send({
					inNearOrLikelyToAffectDesignatedSitesCheckboxes: ['SAC', 'SPA'],
					customDesignation: ''
				});

			expect(mockedLPAQPatchEndpoint.isDone()).toBe(true);
			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/2/lpa-questionnaire/1'
			);
		});

		it('should re-render the change page with the expected error message and not call the PATCH LPA questionnaire endpoint if the "Other" checkbox is checked but custom designation text is not populated in the conditional text input', async () => {
			const mockedLPAQPatchEndpoint = nock('http://test/')
				.patch('/appeals/2/lpa-questionnaires/1')
				.reply(200, {});

			const response = await request
				.post(
					`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
				)
				.send({
					inNearOrLikelyToAffectDesignatedSitesCheckboxes: 'other',
					customDesignation: ''
				});

			const unprettifiedHtml = parseHtml(response.text, {
				skipPrettyPrint: true,
				rootElement: '.govuk-error-summary'
			}).innerHTML;

			expect(mockedLPAQPatchEndpoint.isDone()).toBe(false);
			expect(response.statusCode).toBe(200);
			expect(unprettifiedHtml).toContain('There is a problem</h2>');
			expect(unprettifiedHtml).toContain('Enter details of the other designation(s)</a>');
		});

		it('should call the PATCH LPA questionnaire endpoint with the expected payload and redirect to the LPA Questionnaire page if the "Other" checkbox is checked and custom designation text is populated in the conditional text input', async () => {
			const mockedLPAQPatchEndpoint = nock('http://test/')
				.patch('/appeals/2/lpa-questionnaires/1')
				.reply(200, {});

			const response = await request
				.post(
					`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
				)
				.send({
					inNearOrLikelyToAffectDesignatedSitesCheckboxes: 'other',
					customDesignation: 'test custom designation'
				});

			expect(mockedLPAQPatchEndpoint.isDone()).toBe(true);
			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/2/lpa-questionnaire/1'
			);
		});

		it('should call the PATCH LPA questionnaire endpoint with the expected payload and redirect to the LPA Questionnaire page if the "None" checkbox is checked', async () => {
			const mockedLPAQPatchEndpoint = nock('http://test/')
				.patch('/appeals/2/lpa-questionnaires/1')
				.reply(200, {});

			const response = await request
				.post(
					`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
				)
				.send({
					inNearOrLikelyToAffectDesignatedSitesCheckboxes: 'none',
					customDesignation: ''
				});

			expect(mockedLPAQPatchEndpoint.isDone()).toBe(true);
			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/2/lpa-questionnaire/1'
			);
		});

		it('should call the PATCH LPA questionnaire endpoint with the same payload as if the "None" checkbox were checked and redirect to the LPA Questionnaire page if no checkboxes are checked', async () => {
			const mockedLPAQPatchEndpoint = nock('http://test/')
				.patch('/appeals/2/lpa-questionnaires/1')
				.reply(200, {});

			const response = await request
				.post(
					`${baseUrl}/2/lpa-questionnaire/1/in-near-or-likely-to-affect-designated-sites/change`
				)
				.send({
					customDesignation: ''
				});

			expect(mockedLPAQPatchEndpoint.isDone()).toBe(true);
			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/2/lpa-questionnaire/1'
			);
		});
	});
});
