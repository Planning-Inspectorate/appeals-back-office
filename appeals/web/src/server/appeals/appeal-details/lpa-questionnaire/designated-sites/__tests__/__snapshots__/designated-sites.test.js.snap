// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in-near-or-likely-to-affect-designated-sites GET /in-near-or-likely-to-affect-designated-sites/change should render the change in, near or likely to affect designated sites page with the expected content 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <fieldset class="govuk-fieldset">
                                <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
                                    <h1 class="govuk-fieldset__heading"> Change whether the site is in, partly in, or likely to affect a sensitive area</h1>
                                </legend>
                                <div class="govuk-checkboxes" data-module="govuk-checkboxes">
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="SSSI">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes">SSSI (site of special scientific interest)</label>
                                    </div>
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes-2"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="cSAC">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-2">cSAC (candidate special area of conservation)</label>
                                    </div>
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes-3"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="SAC">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-3">SAC (special area of conservation)</label>
                                    </div>
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes-4"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="pSPA">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-4">pSPA (potential special protection area)</label>
                                    </div>
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes-5"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="SPA">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-5">SPA Ramsar (Ramsar special protection area)</label>
                                    </div>
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes-6"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="other" data-aria-controls="conditional-inNearOrLikelyToAffectDesignatedSitesCheckboxes-6">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-6">Other</label>
                                    </div>
                                    <div class="govuk-checkboxes__conditional govuk-checkboxes__conditional--hidden"
                                    id="conditional-inNearOrLikelyToAffectDesignatedSitesCheckboxes-6">
                                        <label for="custom-designation" class="govuk-label">Other designation(s)</label>
                                        <input class="govuk-input" id="custom-designation"
                                        name="customDesignation" type="text" value="">
                                    </div>
                                    <div class="govuk-checkboxes__divider">or</div>
                                    <div class="govuk-checkboxes__item">
                                        <input class="govuk-checkboxes__input" id="inNearOrLikelyToAffectDesignatedSitesCheckboxes-8"
                                        name="inNearOrLikelyToAffectDesignatedSitesCheckboxes" type="checkbox"
                                        value="none" checked data-behaviour="exclusive">
                                        <label class="govuk-label govuk-checkboxes__label" for="inNearOrLikelyToAffectDesignatedSitesCheckboxes-8">No, it is not in, near or likely to affect any designated sites</label>
                                    </div>
                                </div>
                            </fieldset>
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;
